Support for V2 primitives

edoaltamura opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Introduction
This PR introduces support for V2 Primitives within the `EstimatorQNN` and `SamplerQNN` high-level classes, resolving compatibility issues for users running experiments on Qiskit IBM Runtime real backends. The addition enables transitioning between local simulations with V1 primitives and hardware executions with V2 primitives, meeting the new requirements for Qiskit Runtime ISA circuits.

### Problem description
`EstimatorQNN` and `SamplerQNN` default to Estimator and Sampler from `qiskit.primitives`, which only supports V1 primitives. However, Qiskit IBM Runtime requires V2 primitives (introduced in Qiskit Machine Learning 0.8.0) to run jobs on real quantum backends. Using the V1 primitives results in runtime errors because they are incompatible with Qiskit Runtime’s requirements for structured Primitive Unified Blocks (PUBs) and ISA circuit format.

### Changes in This PR
- _Conditional support for V2 Primitives._ Added a check in `EstimatorQNN` and `SamplerQNN` to detect if a V2 primitive (`EstimatorV2` or `SamplerV2`) from `qiskit_ibm_runtime` is provided. If a V2 primitive is detected, the class will handle it automatically, allowing direct submission of circuits to hardware backends. The output of the jobs (both simulated and hardware) remains the same as before. For instance, where the output used to be quasi-probabilities, now the same quasi-probabilities are computed from the raw counts.

- _Backward compatibility with V1 Primitives:_ Users can continue to use V1 primitives (e.g., `SamplerV1` and `EstimatorV1`) for local simulations with Qiskit Aer. This ensures that existing workflows are unaffected and that users can seamlessly transition to real backends by updating the backend and primitive definitions.
> [!WARNING]
> The V1 Primitives are deprecated as of Qiskit Machine Learning `v.0.8.0` and will no longer be officially supported. V2 Primitives will continue to be supported. V1 Primitives will be removed in Qiskit Machine Learning `v.0.9`. 

- _Support for ISA Circuits:_ These updates ensure that circuits and observables are compatible with Qiskit Runtime’s ISA requirements. Added logic to handle circuit transpilation for ISA compatibility when using V2 primitives, with proper mapping of input and weight parameters, especially for composed feature maps and ansatz circuits.
> [!NOTE]
> This update is expected to allow running Qiskit Machine Learning workloads on IBM quantum hardware using Qiskit IBM Runtime (a version of it that includes V2 Primitives).

### Future considerations
- _Deprecation of V1 Primitives:_ This update serves as a transitional support for V1 primitives, which will be deprecated and removed in version 0.9. Users should prepare to migrate fully to V2 primitives to ensure future compatibility.

- _ISA circuit optimization:_ Further work may be needed to refine circuit transpilation and measurement processes for specific hardware constraints and optimize performance across different backends.

### Details and comments
Fixes https://github.com/qiskit-community/qiskit-machine-learning/issues/742
Fixes https://github.com/qiskit-community/qiskit-machine-learning/issues/786
Fixes https://github.com/qiskit-community/qiskit-machine-learning/issues/810
Fixes https://github.com/qiskit-community/qiskit-machine-learning/issues/839
Fixes https://github.com/qiskit-community/qiskit-machine-learning/issues/840

Auxiliary issues closed:
Fixes https://github.com/OkuyanBoga/hc-qiskit-machine-learning/issues/35

Related issues Algorithms:
https://github.com/qiskit-community/qiskit-algorithms/issues/165
https://github.com/qiskit-community/qiskit-algorithms/issues/194
https://github.com/qiskit-community/qiskit-algorithms/issues/164
https://github.com/qiskit-community/qiskit-algorithms/issues/136
https://github.com/qiskit-community/qiskit-algorithms/issues/137

Related issues - others:
https://github.com/qiskit-community/qiskit-finance/issues/338

edoaltamura commented:
We'll need to add Reno files before the merge.

woodsp-ibm commented:
From above: 

> Deprecation of V1 Primitives: This update serves as a transitional support for V1 primitives, which will be deprecated and removed in version 0.9. Users should prepare to migrate fully to V2 primitives to ensure future compatibility.

V1 primitives in Qiskit and Aer are already deprecated in the latest current versions of these. Do you want to issue a deprecation message at this level from ML if the QNN is provided a V1 estimator to state its use is deprecated for ML and to use V2 instead. Yes I would imagine this will be noted in the reno that is yet to be added but having the messages emitted from the code is normally done too. Yes a user would see a couple of messages in this case, one from ML and one from Qiskit/Aer around deprecation of V1, but maybe that'll just reinforce that the user should really be using V2 versions.

woodsp-ibm commented on qiskit_machine_learning/algorithms/inference/qbayesian.py:
```diff
@@ -62,7 +66,8 @@ def __init__(
         limit: int = 10,
         threshold: float = 0.9,
-        sampler: BaseSampler | None = None,
+        sampler: BaseSampler | BaseSamplerV2 | None = None,
```
Below there is this code

```
        if sampler is None:
            sampler = Sampler()
```

which sets a default instance should the user not provide one. Should we already switch this, and other places that do this, over to using V2 now? If there is some behavorial difference that manifests in a different outcome and it should be left like this for now, maybe it should issue a deprecation warning saying that the default will be changed to V2 instance at the same time the V1 primitive support in ML is removed.

edoaltamura replied:
Just updated with this!

woodsp-ibm commented:
With V1 primitives usage being deprecated are you going to swap the notebooks/tutorials over to using V2 which ideally a user should be using now - I did not see any updated by this PR hence this comment

OkuyanBoga commented:
> With V1 primitives usage being deprecated are you going to swap the notebooks/tutorials over to using V2 which ideally a user should be using now - I did not see any updated by this PR hence this comment

Hi Steve, we are planning to swap the notebooks/tutorials after deprecating V1.

woodsp-ibm commented:
I am not sure I understand exactly

> we are planning to swap the notebooks/tutorials after deprecating V1.

V1 primitives are already deprecated - with the V2 being supported now ideally those notebooks should be using V2. Even if you issue no deprecation messages from ML logically the support for V1 is being deprecated now with 0.8 with the intent to remove it completely in 0.9, Hence the notebooks being changed now for 0.8 to me logically is still after the V1 support is deprecated. Even if its not part of this PR I would think it would be done for 0.8, no?

OkuyanBoga commented:
> I am not sure I understand exactly
> 
> > we are planning to swap the notebooks/tutorials after deprecating V1.
> 
> V1 primitives are already deprecated - with the V2 being supported now ideally those notebooks should be using V2. Even if you issue no deprecation messages from ML logically the support for V1 is being deprecated now with 0.8 with the intent to remove it completely in 0.9, Hence the notebooks being changed now for 0.8 to me logically is still after the V1 support is deprecated. Even if its not part of this PR I would think it would be done for 0.8, no?

Due to time constraints, what about doing it in minor updates through 0,8?

woodsp-ibm commented:
> Due to time constraints, what about doing it in minor updates through 0,8?

Notebooks are part of docs and docs can be republished at any time independent of a release. Hopefully they work right still with the V2 changes - in my mind having them updated would have ensured this. They would get updated and backported and the docs republished - simple as that, no version change needed as such since no new package would get released as they are are not part of what gets published to pypi - just part of the docs.

edoaltamura commented:
> From above:
> 
> > Deprecation of V1 Primitives: This update serves as a transitional support for V1 primitives, which will be deprecated and removed in version 0.9. Users should prepare to migrate fully to V2 primitives to ensure future compatibility.
> 
> V1 primitives in Qiskit and Aer are already deprecated in the latest current versions of these. Do you want to issue a deprecation message at this level from ML if the QNN is provided a V1 estimator to state its use is deprecated for ML and to use V2 instead. Yes I would imagine this will be noted in the reno that is yet to be added but having the messages emitted from the code is normally done too. Yes a user would see a couple of messages in this case, one from ML and one from Qiskit/Aer around deprecation of V1, but maybe that'll just reinforce that the user should really be using V2 versions.

I just added the deprecation warnings in the top-level classes.
